cmd/acr/annotate.go

```diff
@@ -0,0 +1,477 @@
+	cmd.Flags().Uint64Var(&annotateParams.filterTimeout, "filter-timeout-seconds", defaultRegexpMatchTimeoutSeconds, "This limits the evaluation of the regex filter, and will return a timeout error if this duration is exceeded during a single evaluation. If written incorrectly a regexp filter with backtracking can result in an infinite loop")
+	cmd.Flags().StringVar(&annotateParams.artifactType, "artifact-type", "", "The configurable artifact type for an organization")
+	cmd.Flags().StringSliceVarP(&annotateParams.annotations, "annotations", "a", []string{}, "The configurable annotation key value that can be specified one or more times")
+	cmd.Flags().BoolVar(&annotateParams.untagged, "untagged", false, "If the untagged flag is set, all the manifests that do not have any tags associated to them will also be annotated, except if they belong to a manifest list that contains at least one tag")
+	cmd.Flags().BoolVar(&annotateParams.dryRun, "dry-run", false, "If the dry-run flag is set, no manifest or tag will be annotated. The output would be the same as if they were annotated")
+	cmd.Flags().IntVar(&annotateParams.concurrency, "concurrency", defaultPoolSize, annotatedConcurrencyDescription)
+	cmd.Flags().BoolP("help", "h", false, "Print usage")
+	cmd.MarkFlagRequired("filter")
+	cmd.MarkFlagRequired("artifact-type")
+	cmd.MarkFlagRequired("annotations")
+	return cmd
+}
+
+// annotateTags annotates all tags that match the tagFilter string.
+func annotateTags(ctx context.Context,
+	acrClient api.AcrCLIClientInterface,
+	orasClient api.ORASClientInterface,
+	poolSize int,
+	loginURL string,
+	repoName string,
+	artifactType string,
+	annotations []string,
+	tagFilter string,
+	regexpMatchTimeoutSeconds uint64) (int, error) {
+
+	fmt.Printf("Annotating tags for repository: %s\n", repoName)
+
+	tagRegex, err := buildRegexFilter(tagFilter, regexpMatchTimeoutSeconds)
+	if err != nil {
+		return -1, err
+	}
+
+	lastTag := ""
+	annotatedTagsCount := 0
+
+	// In order to only have a limited amount of http requests, an annotator is used that will start goroutines to annotate tags.
+	// pass orasClient instead of acrClient
+	annotator, err := worker.NewAnnotator(poolSize, orasClient, loginURL, repoName, artifactType, annotations)
+	if err != nil {
+		return -1, err
+	}
+
+	for {
+		// GetTagsToAnnotate will return an empty lastTag when there are no more tags.
+		tagsToAnnotate, newLastTag, err := getTagsToAnnotate(ctx, acrClient, repoName, tagRegex, lastTag)
+		if err != nil {
+			return -1, err
+		}
+		lastTag = newLastTag
+		if tagsToAnnotate != nil {
+			count, annotateErr := annotator.AnnotateTags(ctx, tagsToAnnotate)
```

Comment:
You are throttling the annotator through a routine pool but not the getTagsToAnnotate portion which means a registry could receive a ton of tag requests quite quickly. Thats usually discouraged since there is no need to get them all at once. I would suggest putting some limit here.

Reply:
Could you explain what you mean by putting a limit? Just not sure how to implement this as this code is mirrored from the purge command code

Reply:
A simple sleep might be enough as an example. What I mean in general is that AnnotateTags happens asynchronously meaning that the routine continues after that immediately which results in going into the next getTagsToAnnotate instantly. Ideally both the getTagstoAnnotate and the AnnotateTags should be done with some limits.  I would for example have one goroutine to run getTagsToannotate, then use a channel to communicate with a goroutine that will run annotateTags. In practice however, I dont think registries have enough tags for this to be a massive concern as we are restricted to the repo, so maybe adding a simple sleep after some number of requests is enough.